docs/architecture/adr-062-p2p-architecture.md

```diff
@@ -0,0 +1,531 @@
+    // addresses do not need to be normalized in any way (e.g. 0.0.0.0 is
+    // valid), as they should be preprocessed before being advertised.
+    Endpoints() []Endpoint
+}
+```
+
+How the transport configures listening is transport-dependent, and not covered by the interface. This typically happens during transport construction.
+
+#### Endpoints
+
+`Endpoint` represents a transport endpoint (e.g. an IP address and port). A connection always has two endpoints: one at the local node and one at the remote peer. An outbound connection to a remote endpoint is made via a `Dial()` call, and inbound connections to listening endpoints are returned via `Accept()`.
+
+The `Endpoint` struct is:
+
+```go
+// Endpoint represents a transport connection endpoint, either local or remote.
+type Endpoint struct {
+    // Protocol specifies the transport protocol, used by the router to pick a
+    // transport for an endpoint.
+    Protocol Protocol
+
+    // Path is an optional, arbitrary transport-specific path or identifier.
+    Path string
+
+    // IP is an IP address (v4 or v6) to connect to. If set, this defines the
+    // endpoint as a networked endpoint.
+    IP net.IP
+
+    // Port is a network port (either TCP or UDP). If not set, a default port
+    // may be used depending on the protocol.
+    Port uint16
+}
+
+// Protocol identifies a transport protocol.
+type Protocol string
+```
+
+Endpoints are arbitrary transport-specific addresses, but if they are networked they must use IP addresses and thus rely on IP as a fundamental packet routing protocol. This enables policies for address discovery, advertisement, and exchange - for example, a private `192.168.0.0/24` IP address should only be advertised to peers on that IP network, while the public address `8.8.8.8` may be advertised to all peers. Similarly, any port numbers if given must represent TCP and/or UDP port numbers, in order to use [UPnP](https://en.wikipedia.org/wiki/Universal_Plug_and_Play) to autoconfigure e.g. NAT gateways. Note that a concrete design for detection and advertisement of externally-reachable addresses (e.g. via NAT gateways) is out of scope for this ADR, and may be covered in a separate ADR.
+
+Non-networked endpoints (without an IP address) are considered local, and will only be advertised to other peers connecting via the same protocol. For example, an in-memory transport might use `Endpoint{Protocol: "memory", Path: "foo"}` as an address for the node "foo", and this should only be advertised to other nodes using `Protocol: "memory"`.
+
+#### Connections and Streams
+
+A connection represents an established transport connection between two endpoints (and thus two nodes), which can be used to exchange bytes via logically distinct IO streams. Connections are set up either via `Transport.Dial()` (outbound) or `Transport.Accept()` (inbound). The caller is responsible for verifying the remote peer's public key as returned by the connection.
+
```

Comment:
How is authenticate done exactly? Once a connection is established on a transport, what does a peer send and how is it verified?

Reply:
Good question! For now we would have to do whatever `MConnection` does, and try to keep the current scheme, but I'm not really sure what we would want to do long-term. We're thinking about maybe adopting Noise once we've cleaned everything up, which I think makes sense.